src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -139,26 +156,39 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
             }
             else
             {
-                // We're in an end tag helper block.
+                // We're in an end tag block.
 
-                var tagNameScope = _tagStack.Count > 0 ? _tagStack.Peek().TagName : string.Empty;
-
-                // Validate that our end tag helper matches the currently scoped tag helper, if not we
-                // need to error.
+                // Validate that our end tag matches the currently scoped tag, if not we may need to error.
                 if (tagNameScope.Equals(tagName, StringComparison.OrdinalIgnoreCase))
                 {
-                    ValidTagStructure(tagName, tagBlock, context);
+                    // If there are additional end tags required before we can build our block it means we're in a
+                    // situation like this: <myth req="..."><myth></myth></myth> where we're at the inside </myth>.
+                    if (currentTagBlockTracker.OpenMatchingTags > 0)
+                    {
+                        currentTagBlockTracker.OpenMatchingTags--;
+
+                        return false;
+                    }
+
+                    ValidateTagStructure(tagName, tagBlock, context);
 
                     BuildCurrentlyTrackedTagHelperBlock();
                 }
                 else
                 {
+                    // If there are not TagHelperDescriptors associated with the end tag block that also have no 
+                    // required attributes then it means we can't be a TagHelper, bail out.
+                    if (!_provider.GetDescriptors(tagName, attributeNames: Enumerable.Empty<string>()).Any())
+                    {
+                        return false;
+                    }
```

Comment:
Ahh you're right. Looked this over in-depth and it ends up creating an error for `<anotherTagHelper>` because it's not self closing and doesn't have an end-tag. In the case that you have something like

```HTML
<tagWithRequiredAttribute attribute="value">
    <anotherTagHelper>
        </tagWithRequiredAttribute>
    </anotherTagHelper>
```

You get an error on the start `<tagWithRequiredAttribute ..>` stating that it's malformed and needs an end tag or needs to be self-closing.

I'd prefer not to add the error case for the end tag saying that it needs a start tag because that'd mean we'd need to look up the entire tag stack for every end tag we encounter on the page; not to mention the error we show the user is workable and makes sense.